math/mathcore/inc/Math/LFSR.h

```diff
@@ -0,0 +1,153 @@
+
+#include <array>
+#include <bitset>
+#include <vector>
+#include <set>
+#include <cmath>
+#include "TError.h"
+
+/// Pseudo Random Binary Sequence (PRBS) generator namespace with functions based
+/// on linear feedback shift registers (LFSR) with a periodicity of 2^n-1
+///
+/// @note It should NOT be used for general-purpose random number generation or any
+/// statistical study, for those cases see e.g. std::mt19937 instead.
+/// 
+/// The goal is to generate binary bit sequences with the same algorithm as the ones usually implemented
+/// in electronic chips, so that the theoretically expected ones can be compared with the acquired sequences.
+/// 
+/// The main ingredients of a PRBS generator are a monic polynomial of maximum degree \f$n\f$, with coefficients
+/// either 0 or 1, and a <a href="https://www.nayuki.io/page/galois-linear-feedback-shift-register">Galois</a>
+/// linear-feedback shift register with a non-zero seed. When the monic polynomial exponents are chosen appropriately,
+/// the period of the resulting bit sequence (0s and 1s) yields \f$2^n - 1\f$.
+/// 
+/// @sa https://gist.github.com/mattbierner/d6d989bf26a7e54e7135, https://root.cern/doc/master/civetweb_8c_source.html#l06030,
+/// https://cryptography.fandom.com/wiki/Linear_feedback_shift_register, https://www3.advantest.com/documents/11348/33b24c8a-c8cb-40b8-a2a7-37515ba4abc8,
+/// https://www.reddit.com/r/askscience/comments/63a10q/for_prbs3_with_clock_input_on_each_gate_how_can/, https://es.mathworks.com/help/serdes/ref/prbs.html,
+/// https://metacpan.org/pod/Math::PRBS, https://ez.analog.com/data_converters/high-speed_adcs/f/q-a/545335/ad9689-pn9-and-pn23
+
+
+
+namespace ROOT::Math::LFSR  {
+
+   /**
+    * @brief Generate the next pseudo-random bit using the current state of a linear feedback shift register (LFSR) and update it
+    * @tparam k the length of the LFSR, usually also the order of the monic polynomial PRBS-k (last exponent)
+    * @tparam nTaps the number of taps
+    * @param lfsr the current value of the LFSR. Passed by reference, it will be updated with the next value
+    * @param taps the taps that will be XOR-ed to calculate the new bit. They are the exponents of the monic polynomial. Ordering is unimportant. Note that an exponent E in the polynom maps to bit index E-1 in the LFSR.
+    * @param left if true, the direction of the register shift is to the left <<, the newBit is set on lfsr at bit position 0 (right). If false, shift is to the right and the newBit is stored at bit position (k-1)
+    * @return the new random bit
+    * @throw an exception is thrown if taps are out of the range [1, k]
+    * @see https://en.wikipedia.org/wiki/Monic_polynomial, https://en.wikipedia.org/wiki/Linear-feedback_shift_register, https://en.wikipedia.org/wiki/Pseudorandom_binary_sequence
+    */
+   template <size_t k, size_t nTaps>
+   static bool
+   NextLFSR(std::bitset<k>& lfsr, const std::array<uint16_t, nTaps> taps, const bool left = true)
```

Comment:
```suggestion
   NextLFSR(std::bitset<k>& lfsr, std::span<const uint16_t> taps, bool left = true)
```
I would use `std::span<const uint16_t>` here and then check in the beginning of the function if the size is `nTaps`, throwing and exception if not. That would not cause additional overhead: you are already doing the runtime bounds check by using `taps.at(j)` instead of `taps[j]`. That would generalize the interface and leave it up to the user what container type to use (`std::array<uint16_t, nTaps>` would transparently work I think).

Also, it's usually not recommended to add `const` to arguments that are taken by value anyway, because it's useless/confusing info for the reader of the signature. If the callee makes a copy anyway, the caller doesn't care if it treats this copy as constant or not.

Note that `std::span` needs to be imported via `<ROOT/RSpan.hxx>` for C++17 compatibility.

Reply:
Hmm for static-bound-checking (static_assert) I thought it more convenient as array.
I removed the consts. (I know it clutters the reader interface. But sometimes it helps to avoid mistakes when first writing the code hehe.)